foojay-resolver/src/main/kotlin/org/gradle/toolchains/foojay/distributions.kt

```diff
@@ -19,35 +19,51 @@ val vendorAliases = mapOf(
 )
 
 val j9Aliases = mapOf(
-    JvmVendorSpec.ADOPTOPENJDK to "AOJ OpenJ9",
     JvmVendorSpec.IBM to "Semeru",
     JvmVendorSpec.IBM_SEMERU to "Semeru",
-    any() to "Semeru",
+    JvmVendorSpec.ADOPTOPENJDK to "AOJ OpenJ9",
 )
 
 fun match(
         distributions: List<Distribution>,
         vendor: JvmVendorSpec,
         implementation: JvmImplementation,
         version: JavaLanguageVersion
-): Distribution? {
-    if (JvmImplementation.J9 == implementation) {
-        return distributions.firstOrNull { it.name == j9Aliases[vendor] }
+): List<Distribution> = when {
+    JvmImplementation.J9 == implementation -> {
+        if (vendor == any()) {
+            distributions
+                    .filter { j9Aliases.values.contains(it.name) }
+                    .sortedBy { j9Aliases.values.indexOf(it.name) }
+        } else {
+            distributions.filter { it.name == j9Aliases[vendor] }
+        }
     }
-    if (JvmVendorSpec.GRAAL_VM == vendor) {
-        return match(distributions, JvmVendorSpec.matching("Graal VM CE " + version.asInt()))
+    JvmVendorSpec.GRAAL_VM == vendor -> match(distributions, JvmVendorSpec.matching("Graal VM CE " + version.asInt()))
+    else -> match(distributions, vendor)
+}
+
+private fun match(distributions: List<Distribution>, vendor: JvmVendorSpec): List<Distribution> {
+    if (vendor == any()) {
+        return allDistributionsPrecededByWellKnownOnes(distributions)
     }
-    return match(distributions, vendor)
+
+    return findExactMatch(distributions, vendor) ?: findAllMatchingDistributions(distributions, vendor)
 }
 
-fun match(distributions: List<Distribution>, vendor: JvmVendorSpec): Distribution? {
-    val exactMatchByAlias = distributions.firstOrNull { it.name == vendorAliases[vendor] }
-    if (exactMatchByAlias != null) return exactMatchByAlias
+private fun allDistributionsPrecededByWellKnownOnes(distributions: List<Distribution>): List<Distribution> =
+    distributions
+        .sortedByDescending { vendorAliases.values.reversed().indexOf(it.name) }
```

Comment:
I think we should avoid computing the reversed list multiple times by extracting it into a local variable outside of this block.